darts/models/forecasting/forecasting_model.py

```diff
@@ -444,105 +696,202 @@ def historical_forecasts(
             inspect.signature(retrain_func).parameters.keys()
         )
 
-        # prepare the start parameter -> pd.Timestamp
-        start = series.get_timestamp_at_point(start)
+        series = series2seq(series)
+        past_covariates = series2seq(past_covariates)
+        future_covariates = series2seq(future_covariates)
 
-        # build the prediction times in advance (to be able to use tqdm)
-        last_valid_pred_time = self._get_last_prediction_time(
-            series, forecast_horizon, overlap_end
-        )
+        # If the model has never been fitted before using historical_forecasts,
+        # we need to know if it uses past or future covariates. The only possible assumption is that
+        # the user is using the same covariates as he would in the fit method.
+        if self._fit_called is False:
+            if past_covariates is not None:
+                self._uses_past_covariates = True
+            if future_covariates is not None:
+                self._uses_future_covariates = True
+
+        forecasts_list = []
```

Comment:
Couldn't we maybe use a generator (yielding the results one-by-one) rather than keeping everything in memory and blocking the call until everything is computed?